packet.go

```diff
@@ -157,61 +168,99 @@ func (h *Header) Marshal() ([]byte, error) {
 	 * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 	 */
 
-	rawHeaderLength := 12 + (len(h.CSRC) * csrcLength)
-	if h.Extension {
-		rawHeaderLength += 4 + len(h.ExtensionPayload)
-	}
-	rawHeader := make([]byte, rawHeaderLength)
+	// Get the initial size of the buffer.
+	origLen := len(buf)
 
-	rawHeader[0] |= h.Version << versionShift
+	b0 := (h.Version << versionShift) | uint8(len(h.CSRC))
 	if h.Padding {
-		rawHeader[0] |= 1 << paddingShift
+		b0 |= 1 << paddingShift
 	}
+
 	if h.Extension {
-		rawHeader[0] |= 1 << extensionShift
+		b0 |= 1 << extensionShift
 	}
-	rawHeader[0] |= uint8(len(h.CSRC))
 
+	b1 := h.PayloadType
 	if h.Marker {
-		rawHeader[1] |= 1 << markerShift
+		b1 |= 1 << markerShift
 	}
-	rawHeader[1] |= h.PayloadType
 
-	binary.BigEndian.PutUint16(rawHeader[seqNumOffset:], h.SequenceNumber)
-	binary.BigEndian.PutUint32(rawHeader[timestampOffset:], h.Timestamp)
-	binary.BigEndian.PutUint32(rawHeader[ssrcOffset:], h.SSRC)
-
-	for i, csrc := range h.CSRC {
-		binary.BigEndian.PutUint32(rawHeader[csrcOffset+(i*csrcLength):], csrc)
+	buf = append(buf,
```

Comment:
If there was a bug in here (not saying there is) I would have a really hard time fixing/maintaining this.

Is it not possible to keep the explicit old style of doing things?

Reply:
The old style requires calling `MarshalSize()` again and expanding the buffer if needed, which I guess doesn't actually take much time. It also means keeping track of the offset in a bunch of places.

I actually think this code is way more readable and verifiable. It removes the slice offsetting and just deals with a byte at a time. The encoding might seem scary, but it's how [binary.BigEndian](https://golang.org/src/encoding/binary/binary.go#L115) works. I think `rawHeader[csrcOffset+(i*csrcLength):]` is waaaay scarier.

I can change it but I'd like you to reconsider. :D

Reply:
The nice thing about using binary.BigEndian is I don't have to worry about a typo, when reviewing this code I have to think about it a lot more. What you are adding now works, but in the future someone might come and make a mistake and it will be harder to catch.

I agree that working with offsets is frustrating also, it would be nice to find the best mix of both worlds.

I would add @backkem @wdouglass @maxhawkins or @kc5nra they all are good bit twiddlers and happy to be overridden if majority disagrees!

Reply:
Doesn't this code fail on a big-endian machine?

Reply:
I'm agnostic here. If we have good enough tests either should be OK.

The append way means we have to keep track of the offset in fewer places which is a plus. The `binary` package way makes it harder to mix up little endian and big endian.

Reply:
This looks ok to me, the shifts should make it endian agnostic. I agree that the big append is less scary then the offset additions.

If there's a real performance benefit then that's enough for me, I vote merge it.

Reply:
P.s. I agree with @maxhawkins... Make sure this is covered in good tests!

Reply:
The performance difference between using `append` and `binary.BigEndian` is minimal to non-existent. Really I just wanted to avoid manually keeping track of offsets, instead relying on `append` to keep track of the offset aka the slice `len`.

If you guys think the shifts are scarier than the old code with `offset += <math>` all over the place, then yeah I can do it the other way.

Reply:
Maybe we can avoid the shifts *and* the offset math by using bytes.Buffer with binary.BigEndian?

Reply:
It's going to be a lot slower, especially because `binary.BigEndian.Write` uses reflection.

I pushed a commit that uses `binary.BigEndian.PutX`. It's 12% slower than the append version and I think the code is more involved. Using the `offset` const values instead of `offset += ` has  basically zero performance difference.

Reply:
Honestly, I think the append implementation is worth the 12% speedup in return for a couple of ugly shifts. If this were C i'd hide the shifts behind a macro, but this isn't C.

Give me a few minutes to review the rest of the PR; i'll let you choose whether to merge the append version or the new PutX version.